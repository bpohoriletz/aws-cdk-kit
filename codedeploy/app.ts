import * as cdk from 'aws-cdk-lib';
import * as codedeploy from 'aws-cdk-lib/aws-codedeploy';

import * as con from '../utils/naming';

export function createCdApplication(resourceNamePrefix: string[], stack: cdk.Stack): codedeploy.ServerApplication {
  const resourceName = con.codedeployApplicationName(resourceNamePrefix);

  const app = new codedeploy.ServerApplication(stack, resourceName, {
    applicationName: resourceName,
  });

  return app;
}
