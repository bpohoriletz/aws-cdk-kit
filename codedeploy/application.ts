import { Stack } from "aws-cdk-lib";
import { Role } from "aws-cdk-lib/aws-iam";
import { ServerApplication } from "aws-cdk-lib/aws-codedeploy";
import * as con from "../naming/resources"

export function createApplication(resourceNamePrefix: string[], appRole: Role, stack: Stack) : ServerApplication {
  const resourceName: string = con.applicationName(resourceNamePrefix);

  const app = new ServerApplication(stack, resourceName, {
    applicationName: resourceName
  });

  return app;
}
