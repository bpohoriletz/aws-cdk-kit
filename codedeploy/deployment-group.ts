import * as cdk from 'aws-cdk-lib';
import * as autoscaling from 'aws-cdk-lib/aws-autoscaling';
import * as codedeploy from 'aws-cdk-lib/aws-codedeploy';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as iam from 'aws-cdk-lib/aws-iam';

import SecurityGroupDirector from '../directors/security-group';
import SecurityGroupBuilder from '../ec2/security-group-builder';
import * as names from '../utils/naming';
import ServerDeploymentGroupDirector from '../directors/server-deployment-group';
import AsgDeploymentGroupBuilder from './server-deployment-group-builders/asg-deployment-group-builder';

export function createPublicDeploymentGroup(
  resourceNamePrefix: string[],
  instanceTypes: ec2.InstanceType[],
  app: codedeploy.ServerApplication,
  instanceProfile: iam.InstanceProfile,
  codedeployRole: iam.Role,
  vpc: ec2.IVpc,
  stack: cdk.Stack
): [codedeploy.ServerDeploymentGroup, ec2.SecurityGroup, autoscaling.AutoScalingGroup] {
  let resourceName = names.ec2SecurityGroupName(resourceNamePrefix, 'asg');
  const securityGroup = new SecurityGroupDirector(SecurityGroupBuilder).constructWebSecurityGroup(
    stack,
    'WebSecurityGroup',
    vpc
  );
  // Define the launch template for Spot instances
  const userData = ec2.UserData.forLinux();
  userData.addCommands('echo Hello World');
  resourceName = names.launchTemplateName(resourceNamePrefix);
  const launchTemplate = new ec2.LaunchTemplate(stack, resourceName, {
    associatePublicIpAddress: true,
    instanceProfile: instanceProfile,
    instanceType: instanceTypes[0],
    launchTemplateName: resourceName,
    machineImage: ec2.MachineImage.latestAmazonLinux2023(),
    securityGroup: securityGroup,
    userData: userData,
  });

  // Create an Auto Scaling Group with MixedInstancesPolicy (Spot + On-Demand)
  resourceName = names.autoscalingGroupName(resourceNamePrefix);
  const autoscalingGroup = new autoscaling.AutoScalingGroup(stack, resourceName, {
    autoScalingGroupName: resourceName,
    vpc: vpc,
    vpcSubnets: { subnetType: ec2.SubnetType.PUBLIC },
    mixedInstancesPolicy: {
      instancesDistribution: {
        onDemandBaseCapacity: 0, // All instances are Spot
        onDemandPercentageAboveBaseCapacity: 0, // 100% Spot instances
      },
      launchTemplate: launchTemplate,
      launchTemplateOverrides: instanceTypes.map((instanceType) => ({
        instanceType: instanceType,
      })),
    },
    minCapacity: 1,
    maxCapacity: 1,
    desiredCapacity: 1,
    healthChecks: autoscaling.HealthChecks.ec2({
      gracePeriod: cdk.Duration.minutes(5),
    }),
  });

  // Create deployment group
  resourceName = names.codedeployDeploymentGroupName(resourceNamePrefix);
  const deplGroup = new codedeploy.ServerDeploymentGroup(stack, resourceName, {
    application: app,
    autoScalingGroups: [autoscalingGroup],
    deploymentGroupName: resourceName,
    deploymentConfig: codedeploy.ServerDeploymentConfig.ALL_AT_ONCE,
    installAgent: true,
    role: codedeployRole,
  });

  return [deplGroup, securityGroup, autoscalingGroup];
}
export function createDeploymentGroupToAsg(
  resourceNamePrefix: string[],
  instanceTypes: ec2.InstanceType[],
  app: codedeploy.ServerApplication,
  instanceProfile: iam.InstanceProfile,
  codedeployRole: iam.Role,
  vpc: ec2.IVpc,
  stack: cdk.Stack
): [codedeploy.ServerDeploymentGroup, ec2.SecurityGroup, autoscaling.AutoScalingGroup] {
  let resourceName = names.ec2SecurityGroupName(resourceNamePrefix, 'asg');
  const deplGroupDirector = new ServerDeploymentGroupDirector(AsgDeploymentGroupBuilder);
  deplGroupDirector.application = app;
  deplGroupDirector.role = codedeployRole;
  const securityGroup = new SecurityGroupDirector(SecurityGroupBuilder).constructWebSecurityGroup(
    stack,
    'WebSecurityGroup',
    vpc
  );
  // Define the launch template for Spot instances
  const userData = ec2.UserData.forLinux();
  userData.addCommands('echo Hello World');
  resourceName = names.launchTemplateName(resourceNamePrefix);
  const launchTemplate = new ec2.LaunchTemplate(stack, resourceName, {
    associatePublicIpAddress: false,
    instanceProfile: instanceProfile,
    instanceType: instanceTypes[0],
    launchTemplateName: resourceName,
    machineImage: ec2.MachineImage.latestAmazonLinux2023(),
    securityGroup: securityGroup,
    userData: userData,
  });

  // Create an Auto Scaling Group with MixedInstancesPolicy (Spot + On-Demand)
  resourceName = names.autoscalingGroupName(resourceNamePrefix);
  const autoscalingGroup = new autoscaling.AutoScalingGroup(stack, resourceName, {
    autoScalingGroupName: resourceName,
    vpc: vpc,
    vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS },
    mixedInstancesPolicy: {
      instancesDistribution: {
        onDemandBaseCapacity: 0, // All instances are Spot
        onDemandPercentageAboveBaseCapacity: 0, // 100% Spot instances
      },
      launchTemplate: launchTemplate,
      launchTemplateOverrides: instanceTypes.map((instanceType) => ({
        instanceType: instanceType,
      })),
    },
    minCapacity: 1,
    maxCapacity: 1,
    desiredCapacity: 1,
    healthChecks: autoscaling.HealthChecks.ec2({
      gracePeriod: cdk.Duration.minutes(5),
    }),
  });
  // Add lifecycle hooks
  autoscalingGroup.addLifecycleHook(names.autoscalingGroupHookName(resourceNamePrefix, 'term'), {
    defaultResult: autoscaling.DefaultResult.CONTINUE,
    heartbeatTimeout: cdk.Duration.minutes(1),
    lifecycleTransition: autoscaling.LifecycleTransition.INSTANCE_TERMINATING,
  });
  autoscalingGroup.addLifecycleHook(names.autoscalingGroupHookName(resourceNamePrefix, 'lnch'), {
    defaultResult: autoscaling.DefaultResult.CONTINUE,
    heartbeatTimeout: cdk.Duration.minutes(1),
    lifecycleTransition: autoscaling.LifecycleTransition.INSTANCE_LAUNCHING,
  });

  // Create deployment group
  resourceName = names.codedeployDeploymentGroupName(resourceNamePrefix);
  deplGroupDirector.autoScalingGroups = [autoscalingGroup];
  const deplGroup = deplGroupDirector.constructAsgGroup(stack, resourceName, resourceName);

  return [deplGroup, securityGroup, autoscalingGroup];
}
export function createDeploymentGroupToEc2(
  resourceNamePrefix: string[],
  server: ec2.Instance,
  app: codedeploy.ServerApplication,
  stackName: string,
  codedeployRole: iam.Role,
  stack: cdk.Stack
): codedeploy.ServerDeploymentGroup {
  const resourceName = names.codedeployDeploymentGroupName(resourceNamePrefix);
  // Tag instance and create tag set
  cdk.Tags.of(server).add('Environment', stackName);
  cdk.Tags.of(server).add('Application', app.applicationName);
  const tagSet = new codedeploy.InstanceTagSet(
    {
      Environment: [stackName],
    },
    {
      Application: [app.applicationName],
    }
  );
  // Create deployment group
  const deplGroup = new codedeploy.ServerDeploymentGroup(stack, resourceName, {
    application: app,
    deploymentGroupName: resourceName,
    deploymentConfig: codedeploy.ServerDeploymentConfig.ALL_AT_ONCE,
    ec2InstanceTags: tagSet,
    role: codedeployRole,
  });

  return deplGroup;
}
