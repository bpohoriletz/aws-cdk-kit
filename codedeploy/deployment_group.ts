import * as cdk from "aws-cdk-lib";
import * as autoscaling from "aws-cdk-lib/aws-autoscaling";
import * as codedeploy from "aws-cdk-lib/aws-codedeploy";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as iam from "aws-cdk-lib/aws-iam";

import * as names from "../utils/naming";

export function createDeploymentGroupToAsg(
  resourceNamePrefix: string[],
  instanceTypes: ec2.InstanceType[],
  app: codedeploy.ServerApplication,
  stackName: string,
  instanceProfile: iam.InstanceProfile,
  codedeployRole: iam.Role,
  vpc: ec2.IVpc,
  stack: cdk.Stack): [codedeploy.ServerDeploymentGroup, ec2.SecurityGroup, autoscaling.AutoScalingGroup] {

    let resourceName = names.ec2SecurityGroupName(resourceNamePrefix, "asg");
    const securityGroup = createWebSecurityGroup(resourceNamePrefix, vpc, stack);
    // Define the launch template for Spot instances
    resourceName = names.launchTemplateName(resourceNamePrefix);
    const launchTemplate = new ec2.LaunchTemplate(stack, resourceName, {
      associatePublicIpAddress: false,
      instanceProfile: instanceProfile,
      instanceType: instanceTypes[0],
      launchTemplateName: resourceName,
      machineImage: ec2.MachineImage.latestAmazonLinux2023(),
      securityGroup: securityGroup,
    });

    // Create an Auto Scaling Group with MixedInstancesPolicy (Spot + On-Demand)
    resourceName = names.autoscalingGroupName(resourceNamePrefix);
    const autoscalingGroup = new autoscaling.AutoScalingGroup(stack, resourceName, {
      autoScalingGroupName: resourceName,
      vpc: vpc,
      vpcSubnets: { subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS },
      mixedInstancesPolicy: {
        instancesDistribution: {
          onDemandBaseCapacity: 0, // All instances are Spot
          onDemandPercentageAboveBaseCapacity: 0, // 100% Spot instances
        },
        launchTemplate: launchTemplate,
        launchTemplateOverrides: instanceTypes.map((instanceType) => ({
          instanceType: instanceType,
        })),
      },
      minCapacity: 1,
      maxCapacity: 1,
      desiredCapacity: 1,
      healthCheck: autoscaling.HealthCheck.ec2({
        grace: cdk.Duration.minutes(5),
      }),
    });

    // Create deployment group
    resourceName = names.codedeployDeploymentGroupName(resourceNamePrefix);
    const deplGroup = new codedeploy.ServerDeploymentGroup(stack, resourceName, {
      application: app,
      autoScalingGroups: [autoscalingGroup],
      deploymentGroupName: resourceName,
      deploymentConfig: codedeploy.ServerDeploymentConfig.ALL_AT_ONCE,
      installAgent: true,
      role: codedeployRole,
    });

    return [deplGroup, securityGroup, autoscalingGroup];
}
export function createDeploymentGroupToEc2(
  resourceNamePrefix: string[],
  server: ec2.Instance,
  app: codedeploy.ServerApplication,
  stackName: string,
  codedeployRole: iam.Role,
  stack: cdk.Stack): codedeploy.ServerDeploymentGroup {
    const resourceName = names.codedeployDeploymentGroupName(resourceNamePrefix);
    // Tag instance and create tag set
    cdk.Tags.of(server).add("Environment", stackName);
    cdk.Tags.of(server).add("Application", app.applicationName);
    const tagSet = new codedeploy.InstanceTagSet(
      {
        "Environment": [stackName],
      },
      {
        "Application": [app.applicationName],
      }
    );
    // Create deployment group
    const deplGroup = new codedeploy.ServerDeploymentGroup(stack, resourceName, {
      application: app,
      deploymentGroupName: resourceName,
      deploymentConfig: codedeploy.ServerDeploymentConfig.ALL_AT_ONCE,
      ec2InstanceTags: tagSet,
      role: codedeployRole,
    });

    return deplGroup;
}

// TOFIX; import
function createWebSecurityGroup(resourceNamePrefix: string[], vpc: ec2.IVpc, stack: cdk.Stack): ec2.SecurityGroup {
  const resourceName = names.ec2SecurityGroupName(resourceNamePrefix, "ec2");
  const securityGroup = new ec2.SecurityGroup(stack, resourceName , {
    vpc: vpc,
    description: "Security Group for the EC2",
    securityGroupName: resourceName,
    allowAllOutbound: true
  });

  // Allow Security Group inbound traffic for load balancer
  securityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(80), "Allow incoming traffic over port 80");
  securityGroup.addIngressRule(ec2.Peer.anyIpv4(), ec2.Port.tcp(443), "Allow incoming traffic over port 443");

  return securityGroup;
}
