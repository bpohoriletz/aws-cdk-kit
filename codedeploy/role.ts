import * as cdk from "aws-cdk-lib";
import * as iam from "aws-cdk-lib/aws-iam";

import * as con from "../utils/naming";

// TOFIX; delete
export function createCodedeployRole(resourceNamePrefix: string[], regionalCdBucketArn: string,  stack: cdk.Stack) : iam.Role {
  const resourceName = con.roleName(resourceNamePrefix.concat("codedeploy"));

  const role = new iam.Role(stack, resourceName, {
    assumedBy: new iam.ServicePrincipal("codedeploy.amazonaws.com"),
    managedPolicies: [
      iam.ManagedPolicy.fromAwsManagedPolicyName("service-role/AWSCodeDeployRole"),
    ],
    roleName: resourceName,
  });
  const allowEc2UploadToEbRegionalCdBucket: iam.Policy = new iam.Policy(stack, "allowEc2UploadToEbRegionalCdBucket", {
    statements: [
      new iam.PolicyStatement({
        actions: ["s3:PutObject", "s3:ListBucket", "s3:ListBucketVersions", "s3:GetObject", "s3:GetObjectVersion"],
        resources: [regionalCdBucketArn, `${regionalCdBucketArn}/resources/*`],
      }),
    ],
  });
  allowEc2UploadToEbRegionalCdBucket.attachToRole(role);

  return role;
};
