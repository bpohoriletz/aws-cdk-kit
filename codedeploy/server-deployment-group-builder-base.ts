/* eslint-disable @typescript-eslint/no-unused-expressions */
import * as iam from 'aws-cdk-lib/aws-iam';
import * as codedeploy from 'aws-cdk-lib/aws-codedeploy';
import { ServerDeploymentGroupProduct } from '../products/server-deployment-group';

export default abstract class ServerDeploymentGroupBuilderBase {
  protected deploymentGroupProps: ServerDeploymentGroupProduct;

  constructor() {
    this.deploymentGroupProps = new ServerDeploymentGroupProduct();
  }

  setApplication(application?: codedeploy.IServerApplication): this {
    application && (this.deploymentGroupProps.application = application!);

    return this;
  }

  setName(name?: string): this {
    name && (this.deploymentGroupProps.deploymentGroupName = name!);

    return this;
  }

  setRole(role?: iam.IRole): this {
    role && (this.deploymentGroupProps.role = role!);

    return this;
  }

  getResult(): ServerDeploymentGroupProduct {
    return this.deploymentGroupProps;
  }
}
