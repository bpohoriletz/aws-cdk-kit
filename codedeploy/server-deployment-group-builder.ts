import * as iam from 'aws-cdk-lib/aws-iam';
import * as codedeploy from 'aws-cdk-lib/aws-codedeploy';
import { IServerDeploymentGroupBuilder, ServerDeploymentGroupProduct } from '../products/server-deployment-group';

export default class ServerDeploymentGroupBuilder implements IServerDeploymentGroupBuilder {
  protected deploymentGroupProps: ServerDeploymentGroupProduct;

  constructor() {
    this.deploymentGroupProps = new ServerDeploymentGroupProduct();
  }

  setApplication(application: codedeploy.IServerApplication): IServerDeploymentGroupBuilder {
    this.deploymentGroupProps.setApplication(application);

    return this;
  }

  setDeploymentConfig(): IServerDeploymentGroupBuilder {
    return this;
  }

  setName(name: string): IServerDeploymentGroupBuilder {
    this.deploymentGroupProps.setDeploymentGroupName(name);

    return this;
  }

  setRole(role: iam.IRole): IServerDeploymentGroupBuilder {
    this.deploymentGroupProps.setRole(role);

    return this;
  }

  setInstallAgent(): IServerDeploymentGroupBuilder {
    return this;
  }

  getResult(): ServerDeploymentGroupProduct {
    return this.deploymentGroupProps;
  }
}
