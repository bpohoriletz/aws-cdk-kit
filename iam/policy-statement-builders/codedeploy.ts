import { IPolicyStatementBuilder } from '../products/policy-statement';
import PolicyStatementBuilderBase from '../policy-statement-builder-base';

export default class CodedeployPolicyStatementBuilder
  extends PolicyStatementBuilderBase
  implements IPolicyStatementBuilder
{
  setPermissions(): this {
    this.policyStatementProps.addAction('codedeploy:*');

    return this;
  }
}
