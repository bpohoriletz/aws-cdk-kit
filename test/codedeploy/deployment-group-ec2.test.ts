import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { stub } from "../stubs";

import * as func from "../../codedeploy/deployment-group";

describe("CodeDeploy .createDeploymentGroupToEc2()", () => {
  let stack: cdk.Stack;

  beforeEach(() => { stack = new cdk.Stack(); });

  test("matches snapshot", () => {
    func.createDeploymentGroupToEc2([], stub(stack, "ec2.Instance"),
                                    stub(stack, "cd.ServerApplication"), "stack", stub(stack, "iam.Role"), stack);
    expect(Template.fromStack(stack)).toMatchSnapshot();
  });
});
