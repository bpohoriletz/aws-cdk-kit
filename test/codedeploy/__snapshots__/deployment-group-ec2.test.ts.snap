// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeDeploy .createDeploymentGroupToEc2() matches snapshot 1`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs20.x",
      },
      "cn-northwest-1": {
        "value": "nodejs20.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs20.x",
      },
      "us-gov-west-1": {
        "value": "nodejs20.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "4cd2f2411231aa073e609dc99a49b5c5e974252bf9abf0f9350d237d17cce8be.zip",
        },
        "FunctionName": "Stub-ec2-Instance",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "RoleID477B5CA1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "AWS": "*",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSCodeDeployRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ServerApplicationID0A7B7E9A": {
      "Properties": {
        "ComputePlatform": "Server",
      },
      "Type": "AWS::CodeDeploy::Application",
    },
    "Stubec2Instance9DD605F6": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "Stubec2InstanceCustomResourcePolicyD24E1F6C",
      ],
      "Properties": {
        "Create": "{"service":"sts","action":"GetCallerIdentity","physicalResourceId":{"id":"CustomID"}}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
      },
      "Type": "Custom::AWS",
      "UpdateReplacePolicy": "Delete",
    },
    "Stubec2InstanceCustomResourcePolicyD24E1F6C": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sts:GetCallerIdentity",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "Stubec2InstanceCustomResourcePolicyD24E1F6C",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "prefixdg66D052C2": {
      "Properties": {
        "ApplicationName": {
          "Ref": "ServerApplicationID0A7B7E9A",
        },
        "AutoRollbackConfiguration": {
          "Enabled": true,
          "Events": [
            "DEPLOYMENT_FAILURE",
          ],
        },
        "DeploymentConfigName": "CodeDeployDefault.AllAtOnce",
        "DeploymentGroupName": "pre-fix-dg",
        "Ec2TagSet": {
          "Ec2TagSetList": [
            {
              "Ec2TagGroup": [
                {
                  "Key": "Environment",
                  "Type": "KEY_AND_VALUE",
                  "Value": "stack",
                },
              ],
            },
            {
              "Ec2TagGroup": [
                {
                  "Key": "Application",
                  "Type": "KEY_AND_VALUE",
                  "Value": {
                    "Ref": "ServerApplicationID0A7B7E9A",
                  },
                },
              ],
            },
          ],
        },
        "ServiceRoleArn": {
          "Fn::GetAtt": [
            "RoleID477B5CA1",
            "Arn",
          ],
        },
      },
      "Type": "AWS::CodeDeploy::DeploymentGroup",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
